wso2/product-microgateway
# Validate subscription in API key based on "APIM APIkey" issuer

(this PR is based on changes suggested during discussions and review comments for 1.1.0.8)
### Purpose
- Fix certificate alias assignment and sub validation for API key and Internal key
  - Unlike oauth tokens, certs for API key and Internal key are not dynamic. They are configured during Choreo Connect startup. Hence, separating the dependency on the JWT ID in the header when picking the cert (alias). This would avoid having behaviors that are not visible in the config.
- Introduce a new issuer "APIM APIkey"
  ```
  [[enforcer.security.tokenService]]
    # Provide unique name for the JWT issuer
    name = "APIM APIkey"
    validateSubscription = true
    # Alias name given in Enforcer truststore for the public certificate of the JWT issuer
    certificateAlias = "apikey_certificate_alias"
    # Certificate Filepath within Enforcer
    certificateFilePath = "/home/wso2/security/truststore/wso2carbon.pem"
  ```
- For backward compatibility, we would still continue to depend on "APIM Publisher' issuer when "APIM APIkey" issuer is not present. "APIM Publisher" was originally used to refer to the Control Plane while maintaining a config for API key that was separate from the access token issuer (which was also Resident Key Manager).

  - "Resident Key manager" - Access tokens
  - "MGW"                 - Enforcer Test Key
  - "APIM APIkey"     - APIM API key
  - "APIM Publisher" - APIM Internal key (In case if "APIM APIkey" was not present will be used for API keys as well.)

- Access tokens, Internal Keys, API keys can have their own cert and cert alias for JWT signature validation.
- Disabling subscription validation for API key, will enable using same API key for any API as long as the signature is valid.

### Other Implementation details
- Add new param "ControlPlaneEnabled" to config proto. The reason is,
  - This location https://github.com/wso2/product-microgateway/blob/main/enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/keymgt/KeyManagerHolder.java#L108 set subValidation to true when control plane is enabled. 
  - Yet, there are two ways validate subscriptions.
    - when Control Plane enabled (withAPIM) -> check subscription store of enforcer.
    - when Control Plane disabled (standalone) -> check claims against invoked API 
- Update debug log to exclude query params.

Test changes - standalone
- Update existing apikey test to check sub validation failure for different API 
- Add apikey test with config including "APIM APIkey" issuer 
- Add test for oauth to check sub validation failure for different API  
- In CcWithMultipleEnv, remove the additional apictl env for better clarity of the gateway env
- Remove additional default configs from the test config multiple-env-config.toml

Test changes - withapim - none

### Issues
Fixes https://github.com/wso2/product-microgateway/issues/2906
Fixes https://github.com/wso2/product-microgateway/issues/3058

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)
